win32ss/user/user32/windows/cursoricon.c

```diff
@@ -1,17 +1,243 @@
+        png_destroy_read_struct(&png_ptr, NULL, NULL);
+        return NULL;
+    }
+
+    /* Set our own read_function */
+    PNG_READER_STATE reader_state = { pngbits, filesize, PNG_CHECK_SIG_SIZE };
+    png_set_read_fn(png_ptr, &reader_state, read_memory_png);
+    png_set_sig_bytes(png_ptr, PNG_CHECK_SIG_SIZE);
+
+    /* Read png info */
+    png_read_info(png_ptr, info_ptr);
+
+    /* Add translation of some PNG formats and update info */
+    int color_type = png_get_color_type(png_ptr, info_ptr);
+    if (color_type == PNG_COLOR_TYPE_PALETTE)
+        png_set_palette_to_rgb(png_ptr);
+    else if (color_type == PNG_COLOR_TYPE_GRAY || color_type == PNG_COLOR_TYPE_GRAY_ALPHA)
+        png_set_gray_to_rgb(png_ptr);
+    png_set_scale_16(png_ptr); /* Convert 16-bit channels to 8-bit */
+    png_read_update_info(png_ptr, info_ptr);
+
+    /* Get updated png info */
+    png_uint_32 width, height;
+    int bit_depth;
+    png_get_IHDR(png_ptr, info_ptr, &width, &height, &bit_depth, &color_type, NULL, NULL, NULL);
+    TRACE("width %d, height %d, bit_depth %d, color_type %d\n",
+          width, height, bit_depth, color_type);
+
+    int channels = png_get_channels(png_ptr, info_ptr);
+    int rowbytes = png_get_rowbytes(png_ptr, info_ptr);
+    int image_size = height * rowbytes;
+    TRACE("rowbytes %d, channels %d, image_size %d\n", rowbytes, channels, image_size);
+
+    /* Allocate rows data */
+    png_bytepp row_pointers = png_malloc(png_ptr, sizeof(png_bytep) * height);
+    if (!row_pointers)
+    {
+        ERR("png_malloc failed\n");
+        png_destroy_read_struct(&png_ptr, &info_ptr, NULL);
+        return NULL;
+    }
+
+    for (int i = 0; i < (int)height; i++)
+    {
+        row_pointers[i] = png_malloc(png_ptr, rowbytes);
+        if (!row_pointers[i])
+        {
+            ERR("png_malloc failed\n");
+
+            /* Clean up */
+            while (--i >= 0)
+                png_free(png_ptr, row_pointers[i]);
+            png_free(png_ptr, row_pointers);
+            png_destroy_read_struct(&png_ptr, &info_ptr, NULL);
+
+            return NULL;
+        }
+    }
+
+    /* Read png image data */
+    png_set_rows(png_ptr, info_ptr, row_pointers);
+    png_read_image(png_ptr, row_pointers);
+    png_read_end(png_ptr, info_ptr);
+
+    /* After reading the image, you can deal with row pointers */
+    LPBYTE image_bytes = HeapAlloc(GetProcessHeap(), 0, image_size);
+    if (image_bytes)
+    {
+        LPBYTE pb = image_bytes;
+        for (int i = height - 1; i >= 0; i--)
+        {
+            png_bytep row = row_pointers[i];
+            for (int j = 0; j < channels * width; j += channels)
+            {
+                *pb++ = row[j + 2]; /* Red */
+                *pb++ = row[j + 1]; /* Green */
+                *pb++ = row[j + 0]; /* Blue */
+                if (channels == 4)
+                    *pb++ = row[j + 3]; /* Alpha */
+            }
+            pb += (channels * width) % 4;
+        }
```

Comment:
Would this be faster?

```suggestion
        const UINT alpha_offset = (channels == 4) ? 3 : 0;
        LPBYTE pb = image_bytes;
        for (int i = height - 1; i >= 0; i--)
        {
            png_bytep row = row_pointers[i];
            for (int j = 0; j < channels * width; j += channels)
            {
                *pb++ = row[j + alpha_offset]; /* Alpha */
                *pb++ = row[j + 2]; /* Red */
                *pb++ = row[j + 1]; /* Green */
                *pb++ = row[j + 0]; /* Blue */
            }
            pb += (channels * width) % 4;
        }
```